### PR TITLE
refs #741: better save dialogs

### PR DESCRIPTION
**Pull request by incanus**
Some niceties around default file names, extensions, and save paths. 

Goes from: 

79761412-2f07-11e4-8ccb-24e4b96752f4

For every type of tile to more custom: 

screen shot 2014-08-29 at 11 47 52 am

screen shot 2014-08-29 at 11 56 31 am

**TODO:**
- [ ] Test on Linux
- [ ] Test on Windows (particularly for slashes and what `~/` equates to). 

Also, this punts on auto-adding the appropriate file extension if the user leaves it off (which we punt on now). Attempting that leads to silent overwrites of existing files (good times) since you've bypassed the save dialog's built-in de-duping.

### CODE REVIEW

File: app/atom.js [hunk 1]
```diff
@@ -22,25 +22,46 @@ $(document).ready(function() {
...
+            ],
+            'Tiles': [
+                'mbtiles'
+            ],
+            'Image': [
+                'png',
+                'jpg',
+                'jpeg'
+            ]
+        }
```
Comment (yhahn) [inline app/atom.js:31]: Hrm, I would prob express this as:

```
var fileTypes = {
    tm2z: 'Package',
    mbtiles: 'Tiles',
    png: 'Image',
    jpg: 'Image',
    jpeg: 'Image'
};
```

A little dupe in exchange for simpler lookup.

Comment (incanus) [inline app/atom.js:31]: :+1:

**Comment by camilleanne** [comment]
Windows Testing results:
8.1 & 7: The prefilled file name worked, however, the extension is missing. If you save your file as the new default prefilled name, the file extension must be implied because the file system will recognize it's a png or jpeg. However, if you use your own file name, the implied extension must be gone because it's now an unrecognized file type.
screen shot 2014-09-12 at 11 45 43 am

screen shot 2014-09-12 at 12 04 39 pm

**Comment by incanus** [comment]
Thanks for looking into this @camilleanne — on my list to hit Win & Lin this week. 

I think I'm seeing the same thing in OS X, if I understand you right. Note how my Mac screenshots show the blue-highlighted filename like `Untitled Image.png`. If you start typing, the entry replaces the whole selection (including extension) and you get the same effect. It's not great but this is what I meant above by: 

> Also, this punts on auto-adding the appropriate file extension if the user leaves it off (which we punt on now). Attempting that leads to silent overwrites of existing files (good times) since you've bypassed the save dialog's built-in de-duping.

If the user confirms they want to save `myfile` and you've already cleared the built-in (in at least OS X's case) file replace prompt, but then you silently append a `.png` to what they've already approved, you then get into programmatic de-dupe yourself to make sure that you don't silently overwrite an existing `myfile.png`.

**Comment by incanus** [comment]
One way around this (not saying it's needed, but an option) is to tap into platform-specific facilities for this and extend Atom's file save accordingly. For example: 

https://developer.apple.com/library/mac/documentation/Cocoa/Reference/ApplicationKit/Classes/NSSavePanel_Class/Reference/Reference.html#//apple_ref/doc/uid/20000309-BAJBCEFH

`[NSSavePanel setAllowedFileTypes:]`

> If no extension is given by the user, the first item in the `allowedFileTypes` will be used as the extension for the save panel. 

i.e. Platforms have this figured out and we could expose it in the Atom cross-platform API.

**Comment by camilleanne** [comment]
@incanus What you're saying makes sense, however, on Windows there is no pre-filled file extension at all, so it's not clear that if you clear the filename that you need to add '.png'.

**Comment by incanus** [comment]
@camilleanne Atom (editor) doesn't seem to be doing anything fancy with Atom (shell) when it comes to file save dialogs (after some extended spelunking to trace the logic): 
- https://github.com/atom/atom/blob/b6b7ce31a8996bd93c7843f11536ddbb384eaa38/package.json#L20
- https://github.com/atom/grunt-download-atom-shell
- https://github.com/atom/atom/blob/b6b7ce31a8996bd93c7843f11536ddbb384eaa38/menus/darwin.cson#L38
- https://github.com/atom/atom/blob/b6b7ce31a8996bd93c7843f11536ddbb384eaa38/src/workspace-view.coffee#L162
- https://github.com/atom/atom/blob/b6b7ce31a8996bd93c7843f11536ddbb384eaa38/src/pane.coffee#L461
- https://github.com/atom/atom/blob/b6b7ce31a8996bd93c7843f11536ddbb384eaa38/src/atom.coffee#L453

**Comment by incanus** [comment]
You know about this, right? So I think we have the same dialogs as Atom (editor): 

anigif-1410918602

**Comment by camilleanne** [comment]
screen shot 2014-09-17 at 2 11 41 pm

**Comment by incanus** [comment]
Nice.

**Pull request merged**